File: laba3/DESCipher.cpp
#include "DESCipher.h"

DESCipher::DESCipher(array<Byte> ^key, array<Byte> ^iv)
{
	des = gcnew DESCryptoServiceProvider;
	des->Mode = CipherMode::CBC;
	des->Padding = PaddingMode::Zeros;
	if(key == nullptr)
		des->GenerateKey();
	else
		des->Key = key;
	if(iv == nullptr)
		des->GenerateIV();
	else
		des->IV = iv;
}
array<Byte>^ DESCipher::GetKey()
{
	return des->Key;
}
array<Byte>^ DESCipher::GetIV()
{
	return des->IV;
}
void DESCipher::Generate()
{
	des->GenerateKey();
	des->GenerateIV();
}
array<Byte>^ DESCipher::Encrypt(array<Byte> ^data)
{
	MemoryStream ^fin = gcnew MemoryStream(data, false);
	MemoryStream ^fout = gcnew MemoryStream();
	CryptoStream ^encStream = gcnew CryptoStream(fout, des->CreateEncryptor(des->Key, des->IV), CryptoStreamMode::Write);
	encStream->Write(data, 0, fin->Length);
	encStream->FlushFinalBlock();
	return fout->ToArray();
}
array<Byte>^ DESCipher::Decrypt(array<Byte> ^data)
{
	MemoryStream ^fin = gcnew MemoryStream(data, false);
	MemoryStream ^fout = gcnew MemoryStream();
	CryptoStream ^encStream = gcnew CryptoStream(fout, des->CreateDecryptor(des->Key, des->IV), CryptoStreamMode::Write);
	encStream->Write(data, 0, fin->Length);
	encStream->FlushFinalBlock();
	return fout->ToArray();
}

File: laba4Client/DESCipher.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;

namespace laba4Client
{
    class DESCipher
    {
        private DES des;
        DESCipher(byte[] key, byte[] iv)
        {
            des = new DESCryptoServiceProvider();
            des.Mode = CipherMode.CBC;
            des.Padding = PaddingMode.Zeros;
            if (key == null)
                des.GenerateKey();
            else
                des.Key = key;
            if (iv == null)
                des.GenerateIV();
            else
                des.IV = iv;
        }
        void Generate()
        {
            des.GenerateKey();
            des.GenerateIV();
        }
        byte[] Encrypt(byte[] data)
        {
            MemoryStream fin = new MemoryStream(data, false);
            MemoryStream fout = new MemoryStream();
            CryptoStream encStream = new CryptoStream(fout, des.CreateEncryptor(des.Key, des.IV), CryptoStreamMode.Write);
            encStream.Write(data, 0, (int)fin.Length);
            encStream.FlushFinalBlock();
            return fout.ToArray();
        }
        byte[] Decrypt(byte[] data)
        {
            MemoryStream fin = new MemoryStream(data, false);
            MemoryStream fout = new MemoryStream();
            CryptoStream encStream = new CryptoStream(fout, des.CreateDecryptor(des.Key, des.IV), CryptoStreamMode.Write);
            encStream.Write(data, 0, (int)fin.Length);
            encStream.FlushFinalBlock();
            return fout.ToArray();
        }
    }
}
